Index: code/q3_ui/ui_team.c
===================================================================
--- code/q3_ui/ui_team.c	(revision 345)
+++ code/q3_ui/ui_team.c	(revision 346)
@@ -41,7 +41,6 @@
 #define ID_JOINBLUE		101
 #define ID_JOINGAME		102
 #define ID_SPECTATE		103
-#define ID_HIDE			104
 
 
 typedef struct
@@ -52,7 +51,6 @@
 	menutext_s		joinblue;
 	menutext_s		joingame;
 	menutext_s		spectate;
-	menutext_s		hide;
 
 	int				localClient;
 } teammain_t;
@@ -93,11 +91,6 @@
 		trap_Cmd_ExecuteText( EXEC_APPEND, va("cmd %s spectator\n", teamCmd) );
 		UI_ForceMenuOff();
 		break;
-
-	case ID_HIDE:
-		trap_Cmd_ExecuteText( EXEC_APPEND, va("cmd %s hide\n", teamCmd) );
-		UI_ForceMenuOff();
-		break;
 	}
 }
 
@@ -111,10 +104,7 @@
 	int		y;
 	int		gametype;
 	char	info[MAX_INFO_STRING];
-	uiClientState_t	cs;
 
-	trap_GetClientState( &cs );
-
 	memset( &s_teammain, 0, sizeof(s_teammain) );
 
 	s_teammain.localClient = localClient;
@@ -178,21 +168,6 @@
 	s_teammain.spectate.color            = colorRed;
 	y += 20;
 
-	s_teammain.hide.generic.type     = MTYPE_PTEXT;
-	s_teammain.hide.generic.flags    = QMF_CENTER_JUSTIFY|QMF_PULSEIFFOCUS;
-	s_teammain.hide.generic.id       = ID_HIDE;
-	s_teammain.hide.generic.callback = TeamMain_MenuEvent;
-	s_teammain.hide.generic.x        = 320;
-	s_teammain.hide.generic.y        = y;
-	s_teammain.hide.string           = "HIDE VIEWPORT";
-	s_teammain.hide.style            = UI_CENTER|UI_SMALLFONT;
-	s_teammain.hide.color            = colorRed;
-	y += 20;
-
-	if (UI_NumLocalClients(&cs) <= 1) {
-		s_teammain.hide.generic.flags  |= QMF_GRAYED;
-	}
-
 	trap_GetConfigString(CS_SERVERINFO, info, MAX_INFO_STRING);   
 	gametype = atoi( Info_ValueForKey( info,"g_gametype" ) );
 			      
@@ -217,7 +192,6 @@
 	Menu_AddItem( &s_teammain.menu, (void*) &s_teammain.joinblue );
 	Menu_AddItem( &s_teammain.menu, (void*) &s_teammain.joingame );
 	Menu_AddItem( &s_teammain.menu, (void*) &s_teammain.spectate );
-	Menu_AddItem( &s_teammain.menu, (void*) &s_teammain.hide );
 }
 
 
